File: src/main/java/com/github/alexandrecarlton/idea/settings/applier/impl/build_execution_deployment/compiler/CompilerSettingsApplier.java
package com.github.alexandrecarlton.idea.settings.applier.impl.build_execution_deployment.compiler;

import com.github.alexandrecarlton.idea.settings.applier.api.SettingsApplier;
import com.github.alexandrecarlton.idea.settings.layout.build_execution_deployment.compiler.CompilerSettings;
import com.intellij.compiler.CompilerConfiguration;
import com.intellij.compiler.CompilerWorkspaceConfiguration;

import javax.inject.Inject;

public class CompilerSettingsApplier implements SettingsApplier<CompilerSettings> {
  private final CompilerConfiguration compilerConfiguration;
  private final CompilerWorkspaceConfiguration compilerWorkspaceConfiguration;

  @Inject
  public CompilerSettingsApplier(
      CompilerConfiguration compilerConfiguration,
      CompilerWorkspaceConfiguration compilerWorkspaceConfiguration) {
    this.compilerConfiguration = compilerConfiguration;
    this.compilerWorkspaceConfiguration = compilerWorkspaceConfiguration;
  }

  @Override
  public void apply(CompilerSettings settings) {
    settings.addRuntimeAssertionsForNotnullAnnotatedMethodsAndParameters()
        .ifPresent(compilerConfiguration::setAddNotNullAssertions);
    settings.buildProcessHeapSizeMbytes().ifPresent(compilerConfiguration::setBuildProcessHeapSize);
    settings.compileIndependentModulesInParallel()
        .ifPresent(parallel -> compilerWorkspaceConfiguration.PARALLEL_COMPILATION = parallel);
  }
}

File: src/main/java/com/github/alexandrecarlton/idea/settings/applier/impl/build_execution_deployment/compiler/CompilerSettingsApplier.kt
package com.github.alexandrecarlton.idea.settings.applier.impl.build_execution_deployment.compiler

import com.github.alexandrecarlton.idea.settings.applier.api.SettingsApplier
import com.github.alexandrecarlton.idea.settings.layout.build_execution_deployment.compiler.CompilerSettings
import com.intellij.compiler.CompilerConfiguration
import com.intellij.compiler.CompilerConfigurationImpl
import com.intellij.compiler.CompilerWorkspaceConfiguration
import com.intellij.openapi.diagnostic.Logger
import javax.inject.Inject

class CompilerSettingsApplier @Inject
constructor(
    private val compilerConfiguration: CompilerConfiguration,
    private val compilerWorkspaceConfiguration: CompilerWorkspaceConfiguration
) : SettingsApplier<CompilerSettings> {

    private val LOG = Logger.getInstance(CompilerSettingsApplier::class.java)

    override fun apply(settings: CompilerSettings) {
        settings.addRuntimeAssertionsForNotnullAnnotatedMethodsAndParameters()
            .ifPresent { compilerConfiguration.isAddNotNullAssertions = it }
        settings.buildProcessHeapSizeMbytes()
            .ifPresent { compilerConfiguration.setBuildProcessHeapSize(it) }
        settings.compileIndependentModulesInParallel()
            .ifPresent { compilerWorkspaceConfiguration.PARALLEL_COMPILATION = it }
        settings.rebuildModuleOnDependencyChange()
            .ifPresent { compilerWorkspaceConfiguration.REBUILD_ON_DEPENDENCY_CHANGE = it }
        settings.sharedBuildProcessVmOptions()
            .ifPresent { compilerConfiguration.buildProcessVMOptions = it }

        if (compilerConfiguration is CompilerConfigurationImpl) {
            settings.resourcePatterns().ifPresent { patterns ->
                compilerConfiguration.removeResourceFilePatterns()
                patterns.forEach { compilerConfiguration.addResourceFilePattern(it) }
            }
        } else {
            LOG.warn("Unable to apply compiler resource pattern; injected CompilerConfiguration is not CompilerConfigurationImpl")
        }
    }
}
